# Assignment 10, Part 2: Data Wrangling

*Learn to prepare data for visualization and analytics.*


## Instructions
This tutorial provides step-by-step training divided into numbered sections. The sections often contain embeded exectable code for demonstration.  This tutorial is accompanied by a practice notebook with the exact same name but with a `-Practice` suffix.  Throughout this tutorial sections labeled "Practice Task" are interspersed and indicated with the icon: ![Task](http://icons.iconarchive.com/icons/sbstnblnd/plateau/16/Apps-gnome-info-icon.png).

You should follow the instructions provided in these sections by performing them in the practice notebook.  When the tutorial is completed you can turn in the final practice notebook. 

## Introduction
The purpose of this assignment is to build on Tidy data cleaning by using Python tools to "massage" or "wrangle" data into formats that are most useful for visualization and analytics.

**What is data wrangling?**

> Data wrangling, sometimes referred to as data munging, is the process of transforming and mapping data from one "raw" data form into another format with the intent of making it more appropriate and valuable for a variety of downstream purposes such as analytics. 

- [Data Wangling](https://en.wikipedia.org/wiki/Data_wrangling) *Wikipedia*

Previously, we learned about Tidy rules for reformatting data.  Transforming data into a Tidy dataset is data wrangling.  We will build on Tidy by introducing some other useful Python functions.

---
## 1. Getting Started
As before, we import any needed packages at the top of our notebook. Let's import Numpy, Pandas, Seaborn and matplotlib:

In [1]:
%matplotlib inline

import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

#### Task 1a: Setup

<span style="float:right; margin-left:10px; clear:both;">![Task](http://icons.iconarchive.com/icons/sbstnblnd/plateau/96/Apps-gnome-info-icon.png)
</span>

Import the following packages:
+ `pandas` as `pd`
+ `numpy` as `np`
+ `matplotlib.pyplot` as `plt`
+ `Seaborn` as `sns`

Activate the `%matplotlib inline` magic.

---
## 2. Data Exploration
The first step in any data analytics task is import and exploration of data.  At this point, we have learned all of the steps we need to identify the data columns, their data types, recognize where we have missing values and recognize categorical and numeric variables in the data.   

For this tutorial we will use a dataset named "Abolone" from the [University of California Irvine Machine Learning Repository](https://archive.ics.uci.edu/ml/datasets/Abalone). The datafile is named `abalone.data` and has been provided to you (no need to download it).  The data has 10 "attributes" or variables. The following table describes these 10 variables, their types, and additional details.

<table>
    <tr><th>Name</th><th>Data Type</th><th>Metric</th><th>Description</th></tr>
    <tr><td>Sample ID</td><td>integer</td><td></td><td>A unique number for each sample taken</td></tr>
    <tr><td>Sex</td><td>nominal</td><td></td><td>M = 0, F = 1, and I = 2 (infant)</td></tr>
	<tr><td>Length</td><td>continuous</td><td>mm</td><td>Longest shell measurement</td></tr>
	<tr><td>Diameter</td><td>continuous</td><td>mm</td><td>perpendicular to length</td></tr>
	<tr><td>Height</td><td>continuous</td><td>mm</td><td>with meat in shell</td></tr>
	<tr><td>Whole weight</td><td>continuous</td><td>grams</td><td>whole abalone</td></tr>
	<tr><td>Shucked weight</td><td>continuous</td><td>grams</td><td>weight of meat</td></tr>
	<tr><td>Viscera weight</td><td>continuous</td><td>grams</td><td>gut weight (after bleeding)</td></tr>
	<tr><td>Shell weight</td><td>continuous</td><td>grams</td><td>after being dried</td></tr>
	<tr><td>Rings</td><td>integer</td><td></td><td>+1.5 gives the age in years</td></tr>
</table>

***Note:*** To demonstrate specific techniques of data wrangling, the dataset provided to you was altered: a sample ID column was added, the Sex column contains numeric IDs, and missing values were added as were duplicates.

Place this dataset into a `data` folder within the same directory as this notebook and import the data. This data has no header information, so, we'll provide it when we import the data:

In [22]:
abalone = pd.read_csv('data/abalone.data', header = None)
abalone.columns = ['Sample_ID','Sex', 'Length', 'Diameter', 'Height', 
          'Whole_weight', 'Shucked_weight', 'Viscera_weight', 
          'Shell_weight', 'Rings']
abalone.head()

Sample_ID  Sex  Length  Diameter  Height  Whole_weight  Shucked_weight  \
0       1000    0   0.455     0.365   0.095        0.5140          0.2245   
1       1001    0   0.350     0.265   0.090        0.2255          0.0995   
2       1002    1   0.530     0.420   0.135        0.6770          0.2565   
3       1003    0   0.440     0.365   0.125        0.5160          0.2155   
4       1004    2   0.330     0.255   0.080        0.2050          0.0895   

   Viscera_weight  Shell_weight  Rings  
0          0.1010         0.150     15  
1          0.0485         0.070      7  
2          0.1415         0.210      9  
3          0.1140         0.155     10  
4          0.0395         0.055      7

### 2.1 Exploring Data Types
First, let's explore how Pandas imported the data types:

In [4]:
abalone.dtypes

Sample_ID           int64
Sex                 int64
Length            float64
Diameter          float64
Height            float64
Whole_weight      float64
Shucked_weight    float64
Viscera_weight    float64
Shell_weight      float64
Rings               int64
dtype: object

Other than the first and last columns, all others were imported as `float64` which is a decimal value. The first column was imported as an `object` which is a string, and the last column was imported as a an `integer`.  This looks correct for the data.

Let's get a sense of how big the data is:

In [23]:
abalone.shape

(4186, 10)

Next, we can explore the distribution of numerical data using the `describe` function:

In [6]:
abalone.describe()

Sample_ID          Sex       Length     Diameter       Height  \
count  4186.000000  4186.000000  4184.000000  4185.000000  4186.000000   
mean   3086.700669     0.955805     0.523635     0.407621     0.139406   
std    1206.816977     0.828239     0.120323     0.099446     0.041890   
min    1000.000000     0.000000     0.075000     0.055000     0.000000   
25%    2039.250000     0.000000     0.450000     0.350000     0.115000   
50%    3085.500000     1.000000     0.545000     0.425000     0.140000   
75%    4131.750000     2.000000     0.615000     0.480000     0.165000   
max    5178.000000     2.000000     0.815000     0.650000     1.130000   

       Whole_weight  Shucked_weight  Viscera_weight  Shell_weight        Rings  
count   4185.000000     4186.000000     4185.000000   4186.000000  4186.000000  
mean       0.827486        0.358964        0.180390      0.238535     9.930244  
std        0.490774        0.222266        0.109723      0.139352     3.222740  
min        0.002000        0.001000        0.000500      0.001500     1.000000  
25%        0.440500        0.185125        0.092500      0.130000     8.000000  
50%        0.798000        0.335500        0.170500      0.233000     9.000000  
75%        1.153000        0.501500        0.252500      0.328500    11.000000  
max        2.825500        1.488000        0.760000      1.005000    29.000000

Observe that even though the 'Sex' column was provided as a numeric value, it is actually categorical, with each sex represented as a unique number.  We can explore the categorical data using the `groupby` function, followed by the `size` function.

In [7]:
abalone.groupby(by=['Sex']).size()

Sex
0    1532
1    1307
2    1347
dtype: int64

### 2.2 Finding Missing Values
Before proceeding with any analysis you should know the state of missing values in the dataset.  For most analytics missing values are not supported. Some tools will automatically ignore them but it may be easier, in some cases, to remove them.

First, let's quantify how many missing values we have. The `isna` function will convert the data into `True` or `False` values: `True` if the value is missing:

In [8]:
abalone.isna().head()

Sample_ID    Sex  Length  Diameter  Height  Whole_weight  Shucked_weight  \
0      False  False   False     False   False         False           False   
1      False  False   False     False   False         False           False   
2      False  False   False     False   False         False           False   
3      False  False   False     False   False         False           False   
4      False  False   False     False   False         False           False   

   Viscera_weight  Shell_weight  Rings  
0           False         False  False  
1           False         False  False  
2           False         False  False  
3           False         False  False  
4           False         False  False

We can use the `sum` function to then identify how many missing values we have per column:

In [9]:
abalone.isna().sum()

Sample_ID         0
Sex               0
Length            2
Diameter          1
Height            0
Whole_weight      1
Shucked_weight    0
Viscera_weight    1
Shell_weight      0
Rings             0
dtype: int64

### 2.3 Inspecting Duplicates
Sometimes we may or may not want duplicates in the data. This depends on the expectations of the experiments and the measurements taken. Sometimes duplicates may represent human error in data entry. So, let's look for duplicated data.  We have 4,184 rows, let's see how many unique values per column that we have:

In [10]:
abalone.nunique()

Sample_ID         4179
Sex                  3
Length             134
Diameter           111
Height              51
Whole_weight      2429
Shucked_weight    1515
Viscera_weight     880
Shell_weight       926
Rings               28
dtype: int64

For all of the columns we have fewer that 4,184 values.  For columns like 'Sex' we have 3 unique values, but these repeated values are expected.  The decimal values also have duplicates. The likelihood of seeing the exact same decimal values varies based on the distribution for the variable and the number of decimal values in the measurement.  The number of duplicated values does not seem unordinary.  However, the sample ID should be unique, yet we have 4,177 of them instead of 4,184. This implies we have duplicated samples in the data. 

We can identify then umber of duplicated 'Sample_ID' values are in the data by using the `duplicated` function. 

In [11]:
abalone.duplicated(subset='Sample_ID').sum()

7

We have 7 duplicated rows. Now let's see which rows have duplicated samples:

In [12]:
abalone[abalone.duplicated(subset='Sample_ID', keep= False)]

Sample_ID  Sex  Length  Diameter  Height  Whole_weight  Shucked_weight  \
711        1711    0   0.375     0.300   0.100        0.2465          0.1040   
712        1712    2   0.280     0.205   0.055        0.1135          0.0450   
713        1713    0   0.355     0.265   0.085        0.2010          0.0690   
714        1714    0   0.350     0.255   0.080        0.1915          0.0800   
715        1715    2   0.275     0.200   0.065        0.1035          0.0475   
716        1716    2   0.290     0.205   0.070        0.0975          0.0360   
717        1717    2   0.250     0.190   0.060        0.0765          0.0360   
4179       1711    0   0.375     0.300   0.100        0.2465          0.1040   
4180       1712    2   0.280     0.205   0.055        0.1135          0.0450   
4181       1713    0   0.355     0.265   0.085        0.2010          0.0690   
4182       1714    0   0.350     0.255   0.080        0.1915          0.0800   
4183       1715    2   0.275     0.200   0.065        0.1035          0.0475   
4184       1716    2   0.290     0.205   0.070        0.0975          0.0360   
4185       1717    2   0.250     0.190   0.060        0.0765          0.0360   

      Viscera_weight  Shell_weight  Rings  
711           0.0475        0.0830     11  
712           0.0275        0.0335      7  
713           0.0530        0.0695      8  
714           0.0385        0.0630      9  
715           0.0205        0.0300      7  
716           0.0190        0.0350      8  
717           0.0115        0.0245      6  
4179          0.0475        0.0830     11  
4180          0.0275        0.0335      7  
4181          0.0530        0.0695      8  
4182          0.0385        0.0630      9  
4183          0.0205        0.0300      7  
4184          0.0190        0.0350      8  
4185          0.0115        0.0245      6

It looks like the rows are exact duplicates, so this was probably human entry error. We need to remove the copies rows. We will do so in the **3.1 Filtering** section below.

---
## 3. Cleanup
### 3.1 Correcting Data Types
During the data exploration phase above, we noticed that the Sex column was provided as a nuumber to represent the Sex category, and therefore, Pandas imported that column as a numeric value. We need to convert that to a categorical value, because the meaning of the column is not ordinal or numeric we should covert it to a string object.

We can do that with two functions that work on Series:  
- `astype`  converts the type of data in the series. 
- `replace`  replaces values in the series.

We'll use `astype` to convert the column to a string and `replace` to convert the numbers to more easily recognizable 'Male', 'Female' and 'Infant' strings.

In [13]:
# First convert the column from an integer to a string.
sex = abalone['Sex'].astype(str)

# Second convert 0 to Male, 1 to Female, and 2 to Infant.
sex = sex.replace('0', 'Male')
sex = sex.replace('1', 'Female')
sex = sex.replace('2', 'Infant')

# Now replace the 'Sex' column of the dataframe with the new Series.
abalone['Sex'] = sex
abalone.head()

Sample_ID     Sex  Length  Diameter  Height  Whole_weight  Shucked_weight  \
0       1000    Male   0.455     0.365   0.095        0.5140          0.2245   
1       1001    Male   0.350     0.265   0.090        0.2255          0.0995   
2       1002  Female   0.530     0.420   0.135        0.6770          0.2565   
3       1003    Male   0.440     0.365   0.125        0.5160          0.2155   
4       1004  Infant   0.330     0.255   0.080        0.2050          0.0895   

   Viscera_weight  Shell_weight  Rings  
0          0.1010         0.150     15  
1          0.0485         0.070      7  
2          0.1415         0.210      9  
3          0.1140         0.155     10  
4          0.0395         0.055      7

In addition, the Sample ID column, despite that it is numeric should not be treated as a numeric column, so let's convert that too:

In [17]:
# Convert Sample_ID to a string
abalone['Sample_ID'] = abalone['Sample_ID'].astype(str)

# Let's check out the datatypes to make sure they match our expectations:
abalone.dtypes

Sample_ID          object
Sex                object
Length            float64
Diameter          float64
Height            float64
Whole_weight      float64
Shucked_weight    float64
Viscera_weight    float64
Shell_weight      float64
Rings               int64
dtype: object

In [18]:
abalone.head()

Sample_ID     Sex  Length  Diameter  Height  Whole_weight  Shucked_weight  \
0      1000    Male   0.455     0.365   0.095        0.5140          0.2245   
1      1001    Male   0.350     0.265   0.090        0.2255          0.0995   
2      1002  Female   0.530     0.420   0.135        0.6770          0.2565   
3      1003    Male   0.440     0.365   0.125        0.5160          0.2155   
4      1004  Infant   0.330     0.255   0.080        0.2050          0.0895   

   Viscera_weight  Shell_weight  Rings  
0          0.1010         0.150     15  
1          0.0485         0.070      7  
2          0.1415         0.210      9  
3          0.1140         0.155     10  
4          0.0395         0.055      7

In [19]:
abalone.describe()

Length     Diameter       Height  Whole_weight  Shucked_weight  \
count  4177.000000  4177.000000  4177.000000   4177.000000     4177.000000   
mean      0.523992     0.407881     0.139516      0.828742        0.359367   
std       0.120093     0.099240     0.041827      0.490389        0.221963   
min       0.075000     0.055000     0.000000      0.002000        0.001000   
25%       0.450000     0.350000     0.115000      0.441500        0.186000   
50%       0.545000     0.425000     0.140000      0.799500        0.336000   
75%       0.615000     0.480000     0.165000      1.153000        0.502000   
max       0.815000     0.650000     1.130000      2.825500        1.488000   

       Viscera_weight  Shell_weight        Rings  
count     4177.000000   4177.000000  4177.000000  
mean         0.180594      0.238831     9.933684  
std          0.109614      0.139203     3.224169  
min          0.000500      0.001500     1.000000  
25%          0.093500      0.130000     8.000000  
50%          0.171000      0.234000     9.000000  
75%          0.253000      0.329000    11.000000  
max          0.760000      1.005000    29.000000

### 3.2 Handling Missing Values
As observed in section 2.2, we do indeed have missing values! Let's remove rows with missing values.  We can do so with the `dropna` function:

In [20]:
abalone = abalone.dropna(axis=0)
abalone.shape

(4177, 10)

Observe that the `axis` argument is set to 0 indicating we will remove rows with missing values. If we compare the `shape` of the dataframe now, with the shape when we first loaded it we will see that we have lost 2 rows with missing values.

In addition to `dropna` you can also use the `fillna` and `replace` functions to rewrite the missing values to something else.

### 3.3 Removing Duplicates

To remove duplicates we can use the [drop_duplicates](http://pandas.pydata.org/pandas-docs/version/0.17/generated/pandas.DataFrame.drop_duplicates.html) function of Pandas.  If we explore the duplicated columns of section 2.3 above we'll see that the rows are the same for all columns.  In this case we can call `drop_duplicates` with no arguments.  However, let's assume we can't guarantee that each column is the same,  but we do want to remove duplicated samples.  We can do this by using the `subset` argument of the `drop_duplicates` function. We don't want to drop all duplicates, we need to keep one set. Therefore, we'll use the `keep` argument to do this.

In [24]:
abalone = abalone.drop_duplicates(['Sample_ID'], keep='first')
abalone.shape

(4179, 10)

In [27]:
  # Finds all rows with sex of "Male" and the number of rings > 3.
matches = (abalone['Sex'] == 'Male') & (abalone['Rings'] > 3)
male = abalone[matches]

TypeError: invalid type comparison

---
## 4. Reshaping Data
Data reshaping is about altering the way data is housed in the data frames of Pandas. It includes filtering of rows, merging data frames, concatenating data frames, grouping, melting and pivoting. We have learned about all of these functions already. As a reminder, the following is a summary of what we've learned:

**Subsetting by Column**:
- *Basic Indexing*
  - Purpose: Allows you to include only columns you need
  - Introduced:  Pandas Part 1 Notebook
  - Example:
  ```python
   # Get the columns: Sample_ID, Sex, Height and Rings
   subset = abalone[['Sample_ID', 'Sex', 'Height', 'Rings']]
  ```
- *`loc` function*
  - Purpose: Allows you to include only columns you need, and performs better for large datasets.
  - Introduced:  Pandas Part 1 Notebook
  - Example:
  ```python
   # Get the columns: Sample_ID, Sex, Height and Rings
   subset = abalone.loc[:,['Sample_ID', 'Sex', 'Height', 'Rings']]
  ```    
  
**Filtering Rows**:
- *Boolean Indexing*
  - Purpose: to filter rows that match desired criteria
  - Introduced:  Pandas Part 1 Notebook
  - Example:  
  ```python
   # Finds all rows with sex of "Male" and the number of rings > 3.
   matches = (abalone['Sex'] == 'Male') & (abalone['Rings'] > 3)
   male = abalone[matches]

   # Or more succinctly
   male = abalone[(abalone['Sex'] == 'Male') & (abalone['Rings'] > 3)]
  ```

**Grouping Data**:
- *`groupby` function*
  - Purpose:  To group rows together that have some similar set of values. Allows you to perform aggregate functions, such as calculating means, summations, sizes. You can create new data frames with aggregated values.
  - Introduced:  Pandas Part 2 Notebook. 
  - Example:
  ```python
  # Calculate the mean column value by each sex:
  abalone.groupby(by="Sex").mean()
  ```
  
**Merging DataFrames**:
- *`concat` function*
  - Purpose: To combine two dataframes.  Depending if the columns and row indexes are the same determines how the data frames are combined.
  - Introduced:  Pandas Part 2 Notebook.

**Melting**:
- *`melt` function*
  - Purpose:  Handles the case where categorical observations are stored in the header labels (i.e. violates Tidy rules).  It moves the header names into a new column and matches the corresponding values.
  - Introduced:  Tidy Part 1 Notebook.

**Pivoting**:
- *`pivot` and `pivot_table` functions*
  - Purpose: The opposite of `melt`. Uses unique values from one more columns to create new columns.
  - Intorduced: Tidy Part 1 Notebook.
  
You can use any of these functions/techniques to reshape the data to meet Tidy standards and appropriate for the analytic or visualization you want to perform.

In practice, the `keep` argument will default to `first` so we don't need to provide it, but including it makes the code more clear.  We have now dropped all duplicated rows and we have 4,177 valid rows

In [ ]:
---
## 4. Reshaping Data
Data reshaping is about altering the way data is housed in the data frames of Pandas. It includes filtering of rows, merging data frames, concatenating data frames, grouping, melting and pivoting. We have learned about all of these functions already. As a reminder, the following is a summary of what we've learned:

**Subsetting by Column**:
- *Basic Indexing*
  - Purpose: Allows you to include only columns you need
  - Introduced:  Pandas Part 1 Notebook
  - Example:
  ```python
   # Get the columns: Sample_ID, Sex, Height and Rings
   subset = abalone[['Sample_ID', 'Sex', 'Height', 'Rings']]
  ```
- *`loc` function*
  - Purpose: Allows you to include only columns you need, and performs better for large datasets.
  - Introduced:  Pandas Part 1 Notebook
  - Example:
  ```python
   # Get the columns: Sample_ID, Sex, Height and Rings
   subset = abalone.loc[:,['Sample_ID', 'Sex', 'Height', 'Rings']]
  ```    
  
**Filtering Rows**:
- *Boolean Indexing*
  - Purpose: to filter rows that match desired criteria
  - Introduced:  Pandas Part 1 Notebook
  - Example:  
  ```python
   # Finds all rows with sex of "Male" and the number of rings > 3.
   matches = (abalone['Sex'] == 'Male') & (abalone['Rings'] > 3)
   male = abalone[matches]

   # Or more succinctly
   male = abalone[(abalone['Sex'] == 'Male') & (abalone['Rings'] > 3)]
  ```

**Grouping Data**:
- *`groupby` function*
  - Purpose:  To group rows together that have some similar set of values. Allows you to perform aggregate functions, such as calculating means, summations, sizes. You can create new data frames with aggregated values.
  - Introduced:  Pandas Part 2 Notebook. 
  - Example:
  ```python
  # Calculate the mean column value by each sex:
  abalone.groupby(by="Sex").mean()
  ```
  
**Merging DataFrames**:
- *`concat` function*
  - Purpose: To combine two dataframes.  Depending if the columns and row indexes are the same determines how the data frames are combined.
  - Introduced:  Pandas Part 2 Notebook.

**Melting**:
- *`melt` function*
  - Purpose:  Handles the case where categorical observations are stored in the header labels (i.e. violates Tidy rules).  It moves the header names into a new column and matches the corresponding values.
  - Introduced:  Tidy Part 1 Notebook.

**Pivoting**:
- *`pivot` and `pivot_table` functions*
  - Purpose: The opposite of `melt`. Uses unique values from one more columns to create new columns.
  - Intorduced: Tidy Part 1 Notebook.
  
You can use any of these functions/techniques to reshape the data to meet Tidy standards and appropriate for the analytic or visualization you want to perform.

In [ ]:
---
## 5. Bring it all together
In order to solidify data wrangling this notebook provides only a single task that attempts to bring together everything you've leanred to this point.

#### Task 5a: Bring it all Together

<span style="float:right; margin-left:10px; clear:both;">![Task](http://icons.iconarchive.com/icons/sbstnblnd/plateau/96/Apps-gnome-info-icon.png)
</span>

You have been provided with two data files:
- `paired_genes.txt`: contains log2 transformed gene expression-levels of two genes from *Oryza sativa* (rice). The data contains 463 data points corresponding to samples of extracted RNA.
- `project_samples.txt`: contains the list of 463 samples and the experimental conditions underwhich the data was collected.  

The `project_samples.txt` file contains 6 variables:  
1. Sample ID
2. Treatment: CONTROL, HEAT, DROUGHT 
3. Time: a time series in minutes 
4. Genotype: one of 4 varities of rice
5. Subspecies: either Indica or  Japonica
6. Subtreatment: CONTORL, HEAT, RECOV_HEAT, DROUGHT or RECOV_DROUGHT

This data was obtained from [NCBI SRA Project PRJNA301554](https://www.ncbi.nlm.nih.gov/bioproject/PRJNA301554/)

Import both of these files and generate a scatterplot comparing the gene expression of the two genes. Color the points in the dataframe by the Subtreatment. Be sure to:
1. Explore the data
2. Clean the data (if needed)
3. Reshape the data by combining both into a single data frame.
4. Plot the scatterplot using Seaborn and the new single data frame.

In [28]:
dfs = pd.read_csv('data/project_samples.txt', sep='\t')
dfs.head(10)

Sample Treatment  Time Genotype Subspecies Subtreatment
0  SRR2931040   CONTROL    15       AZ   Japonica      CONTROL
1  SRR2931041   CONTROL    15       AZ   Japonica      CONTROL
2  SRR2931042   CONTROL    30       AZ   Japonica      CONTROL
3  SRR2931043   CONTROL    30       AZ   Japonica      CONTROL
4  SRR2931044   CONTROL    45       AZ   Japonica      CONTROL
5  SRR2931045   CONTROL    45       AZ   Japonica      CONTROL
6  SRR2931046   CONTROL    60       AZ   Japonica      CONTROL
7  SRR2931047   CONTROL    60       AZ   Japonica      CONTROL
8  SRR2931048   CONTROL    75       AZ   Japonica      CONTROL
9  SRR2931049   CONTROL    75       AZ   Japonica      CONTROL